flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/util/RexProgramProjectExtractor.scala

```diff
@@ -84,6 +105,39 @@ object RexProgramProjectExtractor {
 }
 
 /**
+  * A RexVisitor to extract used nested input fields
+  */
+class RefFieldAccessorVisitor(
+    usedFields: Array[Int],
+    names: List[String])
+  extends RexVisitorImpl[Unit](true) {
+
+  private val group = usedFields.toList
+  private var nestedFields = mutable.LinkedHashSet[String]()
+
+  def getNestedFields: Array[String] = nestedFields.toArray
+
+  override def visitFieldAccess(fieldAccess: RexFieldAccess): Unit = {
+    fieldAccess.getReferenceExpr match {
+      case ref: RexInputRef =>
+        nestedFields += s"${names(ref.getIndex)}.${fieldAccess.getField.getName}"
```

Comment:
Yes, the parent of `RexFieldAccess` can be also a `RexFieldAccess`. We should take care of that.